packages/langium-cli/src/generator/highlighting/monarch-generator.ts

```diff
@@ -0,0 +1,476 @@
+            if(part.start !== '' && part.end !== '' && tokenName === 'comment') {
+                // state-based rule, only add push to jump into it
+                rules.push({
+                    regex: part.start.replace('/', '\\/'),
+                    action: { token: tokenName, next: '@' + tokenName }
+                });
+
+            } else {
+                // single regex rule
+                rules.push({
+                    regex: rule.definition.regex,
+                    action: {token: getMonarchTokenName(rule.name) }
+                });
+            }
+        }
+    }
+    return rules;
+}
+
+/**
+ * Gets comment state rules from the Langium grammar. Accounts for nested multi-line comments.
+ * @param grammar Langium grammar to extract comment rules from
+ * @returns Array of Monarch comment rules
+ */
+function getCommentRules(grammar: langium.Grammar): Rule[] {
+    const rules: Rule[] = [];
+    for(const rule of grammar.rules) {
+        if(isTerminalRule(rule) && (isCommentTerminal(rule) || rule.name === 'WS') && isRegexToken(rule.definition)) {
+            const tokenName = getMonarchTokenName(rule.name);
+            const part = getTerminalParts(terminalRegex(rule))[0];
+            if(part.start !== '' && part.end !== '' && tokenName === 'comment') {
+                // rules to manage comment nesting via push/pop
+                // rule order matters
+
+                const start = part.start.replace('/', '\\/');
+                const end   = part.end.replace('/', '\\/');
+
+                // 1st, add anything that's not in the start sequence
+                rules.push({
+                    regex: `[^${start}]+`,
+                    action: { token: tokenName }
+                });
+
+                // 2nd, otherwise if start seq, push this state again for nesting
+                rules.push({
+                    regex: start,
+                    action: { token: tokenName, next: '@push' }
+                });
+
+                // 3rd, end of sequence, pop this state, keeping others on the stack
+                rules.push({
+                    regex: end,
+                    action: { token: tokenName, next: '@pop' }
+                });
+
+                // 4th, otherwise, start sequence characters are OK in this state
+                rules.push({
+                    regex: `[${start}]`,
+                    action: { token: tokenName }
+                });
+
+            }
+        }
+    }
+    return rules;
+}
+
+/**
+ * Retrieves non-comment terminal rules, creating associated actions for them
+ * @param grammar Grammar to get non-comment terminals from
+ * @returns Array of Rules to add to a Monarch tokenizer state
+ */
+function getTerminalRules(grammar: langium.Grammar): Rule[] {
+    const rules: Rule[] = [];
+    for (const rule of grammar.rules) {
+        if (isTerminalRule(rule) && !isCommentTerminal(rule) && rule.name !== 'WS' && isRegexToken(rule.definition)) {
```

Comment:
Suggestion: Instead of checking for `name === 'WS'` we should rather check whether the terminal regex can match a whitespace character. We do this in other places of the codebase as well.